### PR TITLE
Clarification for only using feature IDs

### PR DESCRIPTION
**Pull request by javagl**
Based on a PR comment, added an explanation and example of using feature IDs without property tables.

(See https://github.com/KhronosGroup/glTF/pull/2082#issuecomment-951415181 )

### CODE REVIEW

**Comment by wallabyway** [comment]
@ptrgags - I like this change.
Connecting to an external database, should be a 'section' in this spec.  It's such an obvious use-case, and one that we will use.

Should I create a PR, with a section, an example and a diagram?

**Comment by javagl** [comment]
Coming back to this after a while, @ptrgags 

> oh another thing, sets of feature IDs may be stored in primitives or nodes in the schema, but conceptually feature IDs themselves are only ever assigned to vertices, GPU instances or texels.
> ...
> hm... in thinking about this some more, "sets" of feature IDs could get a little ambiguous

If I understood that correctly, then the main point was to check whether we say "feature IDs are stored in a mesh primitive" at some point, and change that to something like  *"feature IDs for the vertices are stored in the mesh primitive"*. 

(It may not even be necessary to talk about "sets" here). 

But one could still argue whether they are really _stored_ there, or only _defined_ (or so). This also refers to the suggested change:

> This can also be used for identifying features in a mesh for styling in an application, even if no properties are stored.

But "storing properties" may be a suitable short-hand form for (precise but clumsy) spec-parlance like ~"having non-zero-length `propertyTables` or `propertyTextures` arrays" or so. 

<sub>An aside: If I had wrapped my head about the details of all documents, and all discussions that happened in issue threads, I'd probably have a clear mental model for the vocabularly, when it comes to "defining properties", "storing properties", "storing property values", "storing property definitions" and so on. I think there *are* differences, and I think they matter, but maybe they are less important than I thought.</sub>

---

@wallabyway  

> Connecting to an external database, should be a 'section' in this spec. It's such an obvious use-case, and one that we will use.
> 
> Should I create a PR, with a section, an example and a diagram?

Depending on whether this should be extended to a whole section, I can either extend this PR, or you can make a suggestion. For the diagram (or rather, an image) : We considered to have a somewhat more "generic" image in the overview that shows this case. Right now, the overview shows a table, but this could be replaced by some black box...

Cesium ID Lookup

But there may be ways of representing that in a more suitable form when it refers to a database in particular.

**Comment by wallabyway** [comment]
@javagl - Nice.  That works.

How about a DB icon and something that looks like a DB table ?
(I also had to mention 'floor', because, digital twins 😉)  

141661003-8faf4455-ff06-4ed5-9c10-0514a264dbf7

**Comment by javagl** [comment]
WIth that table, the diagram is again more similar to the original one (before I draw that 'black box' over the table). But I can create a "nice(r)" version of that diagram that can be used in that section, regardless of how far the text is extended or worded exactly.

**Comment by javagl** [comment]
The section "Specifying Feature IDs" now has two subsections, "Resolving Feature IDs with Property Tables" and "Resolving Feature IDs Externally".

Direct preview: https://github.com/javagl/glTF/tree/ext-mesh-features-only-feature-ids/extensions/2.0/Vendor/EXT_mesh_features#specifying-feature-ids

> I think since this example is mainly about showing how to use feature ids without a property table the implicit feature ids might be a bit of a distraction and we should show a more common example using accessors (like "attribute": 0)

On the one hand, I agree that the details for the implicit ID in the "Example" may appear to be distracting. On the other hand, just having an example

> `"EXT_mesh_features": { "featureIds": [  { "attribute": 3 }  ] }`

does not convey much information, and might require a few words as well, and depending on the level of detail, may also be distracting. One could just say: "This defines the attribute that strores the IDs", or  provide some context and say "The `attribute` is used for `FEATURE_ID_<attribute>` that points to an accessor that...".

File: extensions/2.0/Vendor/EXT_mesh_features/README.md [hunk 1]
```diff
@@ -247,7 +247,11 @@ Each feature ID definition may include only a single source, so the following ar
 - `featureId.offset` and `featureId.repeat` (for a Feature ID Attribute)
 - `featureId.index` (for a Feature ID Texture)
 
-Every `propertyTables` index must have an associated `featureIds` definition, but feature IDs may be defined without a property table. The `propertyTables` entry at index `i` corresponds to the `featureIds` entry at the same index. As a result, the length of the `featureIds` array must be greater than or equal to the length of the `propertyTables` array. Each (`featureId`, `propertyTable`) pair must be unique, but individual feature IDs and property tables may be repeated within a primitive or node.
+Feature IDs may be assigned to primitives or nodes
```
Comment (donmccurdy) [inline extensions/2.0/Vendor/EXT_mesh_features/README.md:250]:
```suggestion
Primitives may contain `featureIds` entries for vertex attribute and texture-based feature IDs, and Nodes may contain `featureIds` entries for GPU instance attributes.
```

File: extensions/2.0/Vendor/EXT_mesh_features/README.md [hunk 2]
```diff
@@ -247,7 +247,11 @@ Each feature ID definition may include only a single source, so the following ar
 - `featureId.offset` and `featureId.repeat` (for a Feature ID Attribute)
 - `featureId.index` (for a Feature ID Texture)
 
-Every `propertyTables` index must have an associated `featureIds` definition, but feature IDs may be defined without a property table. The `propertyTables` entry at index `i` corresponds to the `featureIds` entry at the same index. As a result, the length of the `featureIds` array must be greater than or equal to the length of the `propertyTables` array. Each (`featureId`, `propertyTable`) pair must be unique, but individual feature IDs and property tables may be repeated within a primitive or node.
+Feature IDs may be assigned to primitives or nodes 
+
+#### Resolving Feature IDs with Property Tables
```
Comment (donmccurdy) [inline extensions/2.0/Vendor/EXT_mesh_features/README.md:252]:
```suggestion
#### Referencing Property Tables with Feature IDs
```

File: extensions/2.0/Vendor/EXT_mesh_features/README.md [hunk 3]
```diff
@@ -267,6 +271,32 @@ Empty feature IDs (e.g. `{}`) are disallowed — a feature ID must explicitly se
 > }
 > ```
 
+#### Resolving Feature IDs Externally
```
Comment (donmccurdy) [inline extensions/2.0/Vendor/EXT_mesh_features/README.md:274]:
```suggestion
#### Referencing External Resources with Feature IDs
```

File: extensions/2.0/Vendor/EXT_mesh_features/README.md [hunk 4]
```diff
@@ -267,6 +271,32 @@ Empty feature IDs (e.g. `{}`) are disallowed — a feature ID must explicitly se
...
+
+Feature IDs do not have to be associated with a property table. It is also possible to specify feature IDs to only identify the features in an asset, and then use their IDs in an application-specific manner.
```
Comment (donmccurdy) [inline extensions/2.0/Vendor/EXT_mesh_features/README.md:276]:
```suggestion
Feature IDs do not have to be associated with a property table. Without a property table, IDs may identify features for use in other extensions or in custom applications. Use cases for these IDs extend beyond the scope of this extension, but could include identifying features for styling or picking, or looking up metadata externally in a REST API or database.
```

File: extensions/2.0/Vendor/EXT_mesh_features/README.md [hunk 5]
```diff
@@ -267,6 +271,32 @@ Empty feature IDs (e.g. `{}`) are disallowed — a feature ID must explicitly se
...
+> ```jsonc
+> // Primitive:
+> "extensions": {
+>   "EXT_mesh_features": {
+>     "featureIds": [
+>       { "repeat": 3 }
+>     ]
+>   }
+> }
+> ```
```
Comment (donmccurdy) [inline extensions/2.0/Vendor/EXT_mesh_features/README.md:289]: With the illustration below (thanks!), I think it'd be fine to omit the example JSON here. I was going to suggest simplifying to `{"attribute": 0}` but it doesn't really add new information that we haven't seen earlier.

File: extensions/2.0/Vendor/EXT_mesh_features/README.md [hunk 6]
```diff
@@ -267,6 +271,32 @@ Empty feature IDs (e.g. `{}`) are disallowed — a feature ID must explicitly se
...
+
+These feature IDs can be used by the application in different ways. For example they could be used for styling features in a mesh, or to look up information for these features externally, via a REST API, or in a database.
```
Comment (donmccurdy) [inline extensions/2.0/Vendor/EXT_mesh_features/README.md:291]:
```suggestion
```

^Suggest merging this into the previous paragraph; see comment above.

**Review by donmccurdy — COMMENTED** [review]
I think this section is a great addition, thanks! A few suggested wording changes attached.

File: extensions/2.0/Vendor/EXT_mesh_features/README.md (hunk 6 above)
Comment (javagl) [reply extensions/2.0/Vendor/EXT_mesh_features/README.md]: Indeed, the example did not add sooo much value here (and did not really "match" the diagram in any way, FWIW). 
I have removed it, and applied the other changes that you suggested.

File: extensions/2.0/Vendor/EXT_mesh_features/README.md [hunk 7]
```diff
@@ -267,6 +271,17 @@ Empty feature IDs (e.g. `{}`) are disallowed — a feature ID must explicitly se
 > }
 > ```
 
+#### Referencing External Resources with Feature IDs
+
+Feature IDs do not have to be associated with a property table. Without a property table, IDs may identify features for use in other extensions or in custom applications. Use cases for these IDs extend beyond the scope of this extension, but could include identifying features for styling or picking, or looking up metadata externally in a REST API or database. These feature IDs can be used by the application in different ways. For example they could be used for styling features in a mesh, or to look up information for these features externally, via a REST API, or in a database.
```
Comment (donmccurdy) [inline extensions/2.0/Vendor/EXT_mesh_features/README.md:276]:
```suggestion
Feature IDs do not have to be associated with a property table. Without a property table, IDs may identify features for use in other extensions or in custom applications. Use cases for these IDs extend beyond the scope of this extension, but could include identifying features for styling or picking, or looking up metadata externally in a REST API or database.
```

Ended up with two sentences saying the same thing – i think we could remove the last part, or I'm fine with rephrasing this another way!

File: extensions/2.0/Vendor/EXT_mesh_features/README.md [hunk 8]
```diff
@@ -267,6 +271,17 @@ Empty feature IDs (e.g. `{}`) are disallowed — a feature ID must explicitly se
...
+
+> <img src="figures/feature-id-lookup.png"  alt="Feature ID lookup" width="800">
+
+
+
+
+
+
```
Comment (donmccurdy) [inline extensions/2.0/Vendor/EXT_mesh_features/README.md:284]: nit: I think my inline suggestion ended up creating extra empty lines here sorry!

**Review by donmccurdy — COMMENTED** [review]
Two minor comments left, then this looks good to me! The external resources section is not normative presumably, so we can always expand it later if needed.

File: extensions/2.0/Vendor/EXT_mesh_features/README.md (hunk 7 above)
Comment (javagl) [reply extensions/2.0/Vendor/EXT_mesh_features/README.md:276]: This is embarassing. Sorry for not reading the result thoroughly.

(Note to myself: Don't do important commits at 3:00 am....)

**Comment by ptrgags** [comment]
This looks good to me, thanks @javagl!

**Pull request merged**